Extend mhlo::concatenate

As previously filled in #41, we need to extend `mhlo::concaternate` for one of our internally used models. According to @OliverScherf the MLIR snippet

```MLIR
Tensor4D<float, 1, 59, 87, 64> v101 = mhlo::max<>(v100, v74);
Tensor4D<float, 1, 59, 87, 64> v105 = mhlo::max<>(v104, v74);
Tensor4D<float, 1, 59, 87, 128> v106 = mhlo::concatenate<3, Tensor4D<float, 1, 59, 87, 128>>(v101, v105);
```
results in the error message below:
```
/home/oliver/git/mlir-emitc/include/emitc/emitc_mhlo.h: In instantiation of ‘Dest mhlo::concatenate(Src1, Src ...) [with long int Dimension = 3; Dest = Tensor<float, 1, 59, 87, 128>; Src1 = Tensor<float, 1, 59, 87, 64>; Src = {Tensor<float, 1, 59, 87, 64>}]’:
model_generated.h:123:104:   required from here
/home/oliver/git/mlir-emitc/include/emitc/emitc_mhlo.h:458:9: error: no type named ‘type’ in ‘struct concat<3, float, Tensor<float, 1, 59, 87, 64> >’
  458 |   using Rest = typename concat<Dimension, ET_Src, Src...>::type;
      |         ^~~~
/home/oliver/git/mlir-emitc/include/emitc/emitc_mhlo.h: In instantiation of ‘Dest mhlo::concatenate(Src1, Src ...) [with long int Dimension = 3; Dest = Tensor<float, 1, 29, 43, 256>; Src1 = Tensor<float, 1, 29, 43, 128>; Src = {Tensor<float, 1, 29, 43, 128>}]’:
model_generated.h:151:104:   required from here
/home/oliver/git/mlir-emitc/include/emitc/emitc_mhlo.h:458:9: error: no type named ‘type’ in ‘struct concat<3, float, Tensor<float, 1, 29, 43, 128> >’
/home/oliver/git/mlir-emitc/include/emitc/emitc_mhlo.h: In instantiation of ‘Dest mhlo::concatenate(Src1, Src ...) [with long int Dimension = 3; Dest = Tensor<float, 1, 14, 21, 384>; Src1 = Tensor<float, 1, 14, 21, 192>; Src = {Tensor<float, 1, 14, 21, 192>}]’:
model_generated.h:179:104:   required from here
```

--- a/include/emitc/emitc_types.h
+++ b/include/emitc/emitc_types.h
@@ -65,6 +65,30 @@ struct conjunction<B1, Bn...>
 
 template <class... B>
 constexpr bool conjunction_v = conjunction<B...>::value;
+
+// template switch case statement.
+// case_t => condition/case
+template <bool B, typename T>
+struct case_t {
+  static constexpr bool value = B;
+  using type = T;
+};
+
+// switch_t => template switch case
+template <typename First, typename... Rest>
+struct switch_t : std::conditional_t<First::value, First, switch_t<Rest...>> {};
+
+template <typename T>
+struct switch_t<T> {
+  using type = T;
+};
+
+template <bool B, typename T>
+struct switch_t<case_t<B, T>> {
+  // one statement needs to be true
+  static_assert(B, "None of the supplied conditions evaluate to true.");
+  using type = T;
+};
 } // namespace detail
 
 template <typename T, size_t... Shape>
@@ -324,4 +348,48 @@ struct concat<Dim, T, Tensor2D<T, Xs, Ys>...> {
                detail::sum<sizeof...(Ys)>({Ys...})>>;
 };
 
+template <typename T, size_t Dim, size_t... Xs, size_t... Ys, size_t... Zs>
+struct concat<Dim, T, Tensor3D<T, Xs, Ys, Zs>...> {
+  static_assert(0 <= Dim && Dim < 3, "Dimension index out of bounds");
+
+  using type = typename detail::switch_t<
+      detail::case_t<Dim == 0, Tensor3D<T, detail::sum<sizeof...(Xs)>({Xs...}),
+                                        detail::first<sizeof...(Ys)>({Ys...}),
+                                        detail::first<sizeof...(Zs)>({Zs...})>>,
+      detail::case_t<Dim == 1,
+                     Tensor3D<T, detail::first<sizeof...(Xs)>({Xs...}),
+                              detail::sum<sizeof...(Ys)>({Ys...}),
+                              detail::first<sizeof...(Zs)>({Zs...})>>,
+      detail::case_t<Dim == 2,
+                     Tensor3D<T, detail::first<sizeof...(Xs)>({Xs...}),
+                              detail::first<sizeof...(Ys)>({Ys...}),
+                              detail::sum<sizeof...(Zs)>({Zs...})>>>::type;
+};
+
+template <typename T, size_t Dim, size_t... D0, size_t... D1, size_t... D2,
+          size_t... D3>
+struct concat<Dim, T, Tensor4D<T, D0, D1, D2, D3>...> {
+  static_assert(0 <= Dim && Dim < 4, "Dimension index out of bounds");
+
+  using type = typename detail::switch_t<
+      detail::case_t<Dim == 0, Tensor4D<T, detail::sum<sizeof...(D0)>({D0...}),
+                                        detail::first<sizeof...(D1)>({D1...}),
+                                        detail::first<sizeof...(D2)>({D2...}),
+                                        detail::first<sizeof...(D3)>({D3...})>>,
+      detail::case_t<Dim == 1,
+                     Tensor4D<T, detail::first<sizeof...(D0)>({D0...}),
+                              detail::sum<sizeof...(D1)>({D1...}),
+                              detail::first<sizeof...(D2)>({D2...}),
+                              detail::first<sizeof...(D3)>({D3...})>>,
+      detail::case_t<Dim == 2,
+                     Tensor4D<T, detail::first<sizeof...(D0)>({D0...}),
+                              detail::first<sizeof...(D1)>({D1...}),
+                              detail::sum<sizeof...(D2)>({D2...}),
+                              detail::first<sizeof...(D3)>({D3...})>>,
+      detail::case_t<Dim == 3,
+                     Tensor4D<T, detail::first<sizeof...(D0)>({D0...}),
+                              detail::first<sizeof...(D1)>({D1...}),
+                              detail::first<sizeof...(D2)>({D2...}),
+                              detail::sum<sizeof...(D3)>({D3...})>>>::type;
+};
 #endif // EMITC_EMITC_TYPES_H

--- a/unittests/emitc_mhlo.cpp
+++ b/unittests/emitc_mhlo.cpp
@@ -1072,6 +1072,120 @@ TEST(mhlo, concatenate) {
   EXPECT_THAT(lambda_2d_3_col(),
               Pointwise(FloatEq(), {1.0f, 3.0f, 4.0f, 7.0f, 8.0f, 9.0f, 2.0f,
                                     5.0f, 6.0f, 10.0f, 11.0f, 12.0f}));
+
+  Tensor3D<float, 2, 2, 2> t7{1.0f, 2.0f, 3.0f, 4.0f, 5.0f, 6.0f, 7.0f, 8.0f};
+  Tensor3D<float, 2, 2, 2> t8{9.0f,  10.0f, 11.0f, 12.0f,
+                              13.0f, 14.0f, 15.0f, 16.0f};
+  Tensor3D<float, 2, 2, 1> t9{9.0f, 10.0f, 11.0f, 12.0f};
+
+  auto lambda_3d_422 = [&t7, &t8]() -> Tensor3D<float, 4, 2, 2> {
+    return mhlo::concatenate<0, Tensor3D<float, 4, 2, 2>,
+                             Tensor3D<float, 2, 2, 2>,
+                             Tensor3D<float, 2, 2, 2>>(t7, t8);
+  };
+
+  EXPECT_THAT(lambda_3d_422(),
+              Pointwise(FloatEq(),
+                        {1.0f, 2.0f, 3.0f, 4.0f, 5.0f, 6.0f, 7.0f, 8.0f, 9.0f,
+                         10.0f, 11.0f, 12.0f, 13.0f, 14.0f, 15.0f, 16.0f}));
+
+  auto lambda_3d_242 = [&t7, &t8]() -> Tensor3D<float, 2, 4, 2> {
+    return mhlo::concatenate<1, Tensor3D<float, 2, 4, 2>,
+                             Tensor3D<float, 2, 2, 2>,
+                             Tensor3D<float, 2, 2, 2>>(t7, t8);
+  };
+
+  EXPECT_THAT(lambda_3d_242(),
+              Pointwise(FloatEq(),
+                        {1.0f, 2.0f, 3.0f, 4.0f, 9.0f, 10.0f, 11.0f, 12.0f,
+                         5.0f, 6.0f, 7.0f, 8.0f, 13.0f, 14.0f, 15.0f, 16.0f}));
+
+  auto lambda_3d_223 = [&t7, &t9]() -> Tensor3D<float, 2, 2, 3> {
+    return mhlo::concatenate<2, Tensor3D<float, 2, 2, 3>,
+                             Tensor3D<float, 2, 2, 2>,
+                             Tensor3D<float, 2, 2, 1>>(t7, t9);
+  };
+
+  EXPECT_THAT(lambda_3d_223(),
+              Pointwise(FloatEq(), {1.0f, 2.0f, 9.0f, 3.0f, 4.0f, 10.0f, 5.0f,
+                                    6.0f, 11.0f, 7.0f, 8.0f, 12.0f}));
+
+  Tensor4D<float, 2, 2, 2, 2> t10{1.0f,  2.0f,  3.0f,  4.0f,  5.0f,  6.0f,
+                                  7.0f,  8.0f,  9.0f,  10.0f, 11.0f, 12.0f,
+                                  13.0f, 14.0f, 15.0f, 16.0f};
+  Tensor4D<float, 2, 2, 2, 2> t11{17.0f, 18.0f, 19.0f, 20.0f, 21.0f, 22.0f,
+                                  23.0f, 24.0f, 25.0f, 26.0f, 27.0f, 28.0f,
+                                  29.0f, 30.0f, 31.0f, 32.0f};
+  Tensor4D<float, 2, 2, 1, 2> t12{33.0f, 34.0f, 35.0f, 36.0f,
+                                  37.0f, 38.0f, 39.0f, 40.0f};
+
+  auto lambda_4d_4222 = [&t10, &t11]() -> Tensor4D<float, 4, 2, 2, 2> {
+    return mhlo::concatenate<0, Tensor4D<float, 4, 2, 2, 2>,
+                             Tensor4D<float, 2, 2, 2, 2>,
+                             Tensor4D<float, 2, 2, 2, 2>>(t10, t11);
+  };
+
+  EXPECT_THAT(
+      lambda_4d_4222(),
+      Pointwise(FloatEq(),
+                {1.0f,  2.0f,  3.0f,  4.0f,  5.0f,  6.0f,  7.0f,  8.0f,
+                 9.0f,  10.0f, 11.0f, 12.0f, 13.0f, 14.0f, 15.0f, 16.0f,
+                 17.0f, 18.0f, 19.0f, 20.0f, 21.0f, 22.0f, 23.0f, 24.0f,
+                 25.0f, 26.0f, 27.0f, 28.0f, 29.0f, 30.0f, 31.0f, 32.0f}));
+
+  auto lambda_4d_2422 = [&t10, &t11]() -> Tensor4D<float, 2, 4, 2, 2> {
+    return mhlo::concatenate<1, Tensor4D<float, 2, 4, 2, 2>,
+                             Tensor4D<float, 2, 2, 2, 2>,
+                             Tensor4D<float, 2, 2, 2, 2>>(t10, t11);
+  };
+
+  EXPECT_THAT(
+      lambda_4d_2422(),
+      Pointwise(FloatEq(),
+                {1.0f,  2.0f,  3.0f,  4.0f,  5.0f,  6.0f,  7.0f,  8.0f,
+                 17.0f, 18.0f, 19.0f, 20.0f, 21.0f, 22.0f, 23.0f, 24.0f,
+                 9.0f,  10.0f, 11.0f, 12.0f, 13.0f, 14.0f, 15.0f, 16.0f,
+                 25.0f, 26.0f, 27.0f, 28.0f, 29.0f, 30.0f, 31.0f, 32.0f}));
+
+  auto lambda_4d_2242 = [&t10, &t11]() -> Tensor4D<float, 2, 2, 4, 2> {
+    return mhlo::concatenate<2, Tensor4D<float, 2, 2, 4, 2>,
+                             Tensor4D<float, 2, 2, 2, 2>,
+                             Tensor4D<float, 2, 2, 2, 2>>(t10, t11);
+  };
+
+  EXPECT_THAT(
+      lambda_4d_2242(),
+      Pointwise(FloatEq(),
+                {1.0f,  2.0f,  3.0f,  4.0f,  17.0f, 18.0f, 19.0f, 20.0f,
+                 5.0f,  6.0f,  7.0f,  8.0f,  21.0f, 22.0f, 23.0f, 24.0f,
+                 9.0f,  10.0f, 11.0f, 12.0f, 25.0f, 26.0f, 27.0f, 28.0f,
+                 13.0f, 14.0f, 15.0f, 16.0f, 29.0f, 30.0f, 31.0f, 32.0f}));
+
+  auto lambda_4d_2224 = [&t10, &t11]() -> Tensor4D<float, 2, 2, 2, 4> {
+    return mhlo::concatenate<3, Tensor4D<float, 2, 2, 2, 4>,
+                             Tensor4D<float, 2, 2, 2, 2>,
+                             Tensor4D<float, 2, 2, 2, 2>>(t10, t11);
+  };
+
+  EXPECT_THAT(
+      lambda_4d_2224(),
+      Pointwise(FloatEq(),
+                {1.0f,  2.0f,  17.0f, 18.0f, 3.0f,  4.0f,  19.0f, 20.0f,
+                 5.0f,  6.0f,  21.0f, 22.0f, 7.0f,  8.0f,  23.0f, 24.0f,
+                 9.0f,  10.0f, 25.0f, 26.0f, 11.0f, 12.0f, 27.0f, 28.0f,
+                 13.0f, 14.0f, 29.0f, 30.0f, 15.0f, 16.0f, 31.0f, 32.0f}));
+
+  auto lambda_4d_2232 = [&t10, &t12]() -> Tensor4D<float, 2, 2, 3, 2> {
+    return mhlo::concatenate<2, Tensor4D<float, 2, 2, 3, 2>,
+                             Tensor4D<float, 2, 2, 2, 2>,
+                             Tensor4D<float, 2, 2, 1, 2>>(t10, t12);
+  };
+
+  EXPECT_THAT(lambda_4d_2232(),
+              Pointwise(FloatEq(), {1.0f,  2.0f,  3.0f,  4.0f,  33.0f, 34.0f,
+                                    5.0f,  6.0f,  7.0f,  8.0f,  35.0f, 36.0f,
+                                    9.0f,  10.0f, 11.0f, 12.0f, 37.0f, 38.0f,
+                                    13.0f, 14.0f, 15.0f, 16.0f, 39.0f, 40.0f}));
 }
 
 /// Adapted from

--- a/unittests/emitc_types.cpp
+++ b/unittests/emitc_types.cpp
@@ -631,4 +631,39 @@ TEST(types, unravel_index) {
   EXPECT_THAT(t4.unravel_index(211), Pointwise(Eq(), {1, 2, 3, 4}));
 }
 
+TEST(types, concat) {
+  using test_2D_42 =
+      typename concat<0, int, Tensor2D<int, 2, 2>, Tensor2D<int, 2, 2>>::type;
+  using test_2D_24 =
+      typename concat<1, int, Tensor2D<int, 2, 2>, Tensor2D<int, 2, 2>>::type;
+
+  using test_3D_422 = typename concat<0, int, Tensor3D<int, 2, 2, 2>,
+                                      Tensor3D<int, 2, 2, 2>>::type;
+  using test_3D_242 = typename concat<1, int, Tensor3D<int, 2, 2, 2>,
+                                      Tensor3D<int, 2, 2, 2>>::type;
+  using test_3D_224 = typename concat<2, int, Tensor3D<int, 2, 2, 2>,
+                                      Tensor3D<int, 2, 2, 2>>::type;
+
+  using test_4D_4222 = typename concat<0, int, Tensor4D<int, 2, 2, 2, 2>,
+                                       Tensor4D<int, 2, 2, 2, 2>>::type;
+  using test_4D_2422 = typename concat<1, int, Tensor4D<int, 2, 2, 2, 2>,
+                                       Tensor4D<int, 2, 2, 2, 2>>::type;
+  using test_4D_2242 = typename concat<2, int, Tensor4D<int, 2, 2, 2, 2>,
+                                       Tensor4D<int, 2, 2, 2, 2>>::type;
+  using test_4D_2224 = typename concat<3, int, Tensor4D<int, 2, 2, 2, 2>,
+                                       Tensor4D<int, 2, 2, 2, 2>>::type;
+
+  ASSERT_THAT(test_2D_42::shape(), Pointwise(Eq(), {4, 2}));
+  ASSERT_THAT(test_2D_24::shape(), Pointwise(Eq(), {2, 4}));
+
+  ASSERT_THAT(test_3D_422::shape(), Pointwise(Eq(), {4, 2, 2}));
+  ASSERT_THAT(test_3D_242::shape(), Pointwise(Eq(), {2, 4, 2}));
+  ASSERT_THAT(test_3D_224::shape(), Pointwise(Eq(), {2, 2, 4}));
+
+  ASSERT_THAT(test_4D_4222::shape(), Pointwise(Eq(), {4, 2, 2, 2}));
+  ASSERT_THAT(test_4D_2422::shape(), Pointwise(Eq(), {2, 4, 2, 2}));
+  ASSERT_THAT(test_4D_2242::shape(), Pointwise(Eq(), {2, 2, 4, 2}));
+  ASSERT_THAT(test_4D_2224::shape(), Pointwise(Eq(), {2, 2, 2, 4}));
+}
+
 } // namespace

